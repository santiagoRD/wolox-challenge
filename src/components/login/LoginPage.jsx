import React, { useState, lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Loader from '../../helpers/common/Loader';
import useForm from '../../hooks/useForm';
import { validateEmail } from '../../helpers/helpers';
import { userAuthentication, authError } from '../../actions/authActions';
import { BASE_URL } from '../../helpers/constants';
import woloxNavIcon from '../../assets/logo_full_color.svg';
import './LoginPage.scss';

const Button = lazy(() => import('../../helpers/common/Button'));
const Input = lazy(() => import('../../helpers/common/Input'));

const LoginPage = props => {
  const [formValues, handleInputChange] = useForm({
    email: '',
    password: ''
  });
  const [error, setError] = useState({});

  const { email, password } = formValues;
  const { handleLogin, history, handleError, errorRequest } = props;

  const handleSubmit = async e => {
    e.preventDefault();
    if (email && password) {
      try {
        const body = {
          email,
          password
        };
        const {
          data: { token }
        } = await axios.post(`${BASE_URL}/login`, body);
        handleLogin(token, email, true);
        const auth = {
          token,
          email,
          logged: true
        };
        localStorage.setItem('auth', JSON.stringify(auth));
        history.replace('/admin');
      } catch (err) {
        handleError(`Error en el login ${err}`);
      }
    }
  };

  const handleResetError = ({ target }) => {
    setError({
      ...error,
      [target.name]: ''
    });
  };

  const handleValidateEmail = ({ target }) => {
    if (!target.value.trim()) return;
    if (!validateEmail(target.value)) {
      setError({
        ...error,
        email: 'error'
      });
    } else {
      setError({
        ...error,
        email: ''
      });
    }
  };

  return (
    <Suspense fallback={<Loader />}>
      <section className='login__section'>
        <div className='login__header'>
          <Link to='/'>
            <img src={woloxNavIcon} alt='wolox-icon' />
          </Link>
        </div>
        <h2>Bienvenido</h2>
        <form className='form__container' onSubmit={handleSubmit}>
          <div className='form__input__container w-100'>
            <Input
              classList={`${error?.email || ''}`}
              type='email'
              value={email}
              onChange={handleInputChange}
              onBlur={handleValidateEmail}
              onFocus={handleResetError}
              placeholder='Ingresa tu email'
              name='email'
              maxLength={30}
              required
              error={error?.email ? 'Correo ingresado no valido' : ''}
            />
            <Input
              classList={`${error?.passError ? 'error' : ''}`}
              type='password'
              value={password}
              onChange={handleInputChange}
              placeholder='Ingresa tu contraseña'
              name='password'
              required
            />
          </div>
          <div className='form__input__container w-100'>
            <Button type='submit' text='Ingresar' classList='btn btn-bg-blue btn-outline w-100' />
            <Link to='/registry'>
              No tienes cuenta? registrate <span>aquí.</span>
            </Link>
          </div>
        </form>
        {errorRequest && <small className='error'>{errorRequest}</small>}
      </section>
    </Suspense>
  );
};

const mapDispatchToProps = dispatch => bindActionCreators({ handleLogin: userAuthentication, handleError: authError }, dispatch);
const mapStateToProps = ({ error }) => ({ errorRequest: error });

export default connect(mapStateToProps, mapDispatchToProps)(LoginPage);

LoginPage.propTypes = {
  handleLogin: PropTypes.func.isRequired,
  history: PropTypes.shape({
    replace: PropTypes.func
  }).isRequired
};
