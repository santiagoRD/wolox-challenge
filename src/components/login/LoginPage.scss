@import '../../variables.scss';

.login__section {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 300px;
  .login__header {
    position: relative;
    width: 100%;
    a {
      position: absolute;
      width: 150px;
      top: -130px;
    }
  }
  .form__container {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    height: 50vh;
    width: 100%;
    padding: 10px;
    .form__input__container {
      height: 100%;
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: space-evenly;
      a {
        color: $color-text;
        cursor: pointer;
        font-size: $f-size-12;
        transition: all 0.3s ease-in-out;
        &:hover {
          color: $color-blue;
        }
        span {
          text-decoration: underline;
        }
      }
    }
  }
  @media only screen and(min-width: $medium) {
    width: 640px;
    .form__container {
      width: 50%;
    }
  }
}
