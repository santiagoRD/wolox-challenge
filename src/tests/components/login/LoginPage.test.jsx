import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import LoginPage from '../../../components/login/LoginPage';

const middlewares = [];
const mockStore = configureStore(middlewares);

const initialState = {
  user: {
    token: '',
    email: '',
    logged: false
  },
  error: ''
};
let store = mockStore(initialState);
const wrapper = mount(
  <MemoryRouter>
    <Provider store={store}>
      <LoginPage />
    </Provider>
  </MemoryRouter>
);

describe('Testing <LoginPage />', () => {
  beforeEach(() => {
    store = mockStore(initialState);
  });
  test('should render <LoginPage />', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
