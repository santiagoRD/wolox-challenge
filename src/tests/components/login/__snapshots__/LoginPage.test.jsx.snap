// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <LoginPage /> should render <LoginPage /> 1`] = `
<Suspense
  fallback={<Loader />}
>
  <section
    className="login__section"
  >
    <div
      className="login__header"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <img
          alt="wolox-icon"
          src="logo_full_color.svg"
        />
      </a>
    </div>
    <h2>
      Bienvenido
    </h2>
    <form
      className="form__container"
      onSubmit={[Function]}
    >
      <div
        className="form__input__container w-100"
      />
      <div
        className="form__input__container w-100"
      >
        <a
          href="/registry"
          onClick={[Function]}
        >
          No tienes cuenta? registrate 
          <span>
            aqu√≠.
          </span>
        </a>
      </div>
    </form>
  </section>
  <div
    className="loader__wrapper"
  >
    <div
      className="loader__container"
    >
      <div
        className="loader__item"
      />
    </div>
  </div>
</Suspense>
`;
